File: insta360-bridge-flutter/android/app/src/main/kotlin/com/example/insta360_bridge/Insta360Channel.kt
package com.example.insta360_bridge

import android.app.Activity
import android.os.Handler
import android.os.Looper
import android.util.Log
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

class Insta360Channel(
    private val activity: Activity,
    private val methodChannel: MethodChannel
) : MethodCallHandler {

    companion object {
        const val TAG = "Insta360Channel"
    }

    private val mainHandler = Handler(Looper.getMainLooper())
    val wifiHelper = WifiHelper(activity)
    private var demoMode = false
    
    var bgService: Insta360BackgroundService? = null

    fun setService(service: Insta360BackgroundService) {
        this.bgService = service
        Log.d(TAG, "Background Service bound to Flutter Channel!")
    }

    init {
        // Connect WifiHelper results to Flutter events
        wifiHelper.onScanResults = { results ->
            val networkList = results.map { 
                mapOf("ssid" to it.ssid, "level" to it.level)
            }
            invokeDartEvent("onWifiList", networkList)
        }

        wifiHelper.onConnectionStatus = { status ->
            Log.d(TAG, "Wi-Fi connection status: $status")
            invokeDartEvent("onWifiConnected", mapOf("status" to status))
            if (status == "CONNECTED") {
                onWifiConnected()
            }
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "startRecording" -> startRecording(result)
            "stopRecording" -> stopRecording(result)
            "getStatus" -> result.success(bgService?.cameraControl?.getStatus() ?: "DISCONNECTED")
            "connectCamera" -> connectCamera(result)
            "scanWifi" -> scanWifi(result)
            "stopScan" -> {
                wifiHelper.stopScan()
                result.success(null)
            }
            "connectWifi" -> {
                val ssid = call.argument<String>("ssid") ?: ""
                val pwd = call.argument<String>("password") ?: ""
                connectWifi(ssid, pwd, result)
            }
            "saveRecordingMetadata" -> {
                val jsonStr = call.argument<String>("data") ?: ""
                saveRecordingMetadata(jsonStr, result)
            }
            "getRecordings" -> result.success(getRecordings())
            "setDemoMode" -> {
                val enabled = call.argument<Boolean>("enabled") ?: false
                setDemoMode(enabled, result)
            }
            "exportRecording" -> {
                val recordingId = call.argument<String>("recordingId") ?: ""
                exportRecording(recordingId, result)
            }
            "signInToDrive" -> signInToDrive(result)
            "getDriveSignInStatus" -> result.success(getDriveSignInStatus())
            "uploadToDrive" -> {
                val filePath = call.argument<String>("filePath") ?: ""
                uploadToDrive(filePath, result)
            }
            "uploadPendingFiles" -> uploadPendingFiles(result)
            "getPendingUploadCount" -> result.success(bgService?.driveUploader?.getPendingFiles()?.size ?: 0)
            "openWifiSettings" -> {
                val intent = android.content.Intent(android.provider.Settings.ACTION_WIFI_SETTINGS)
                intent.flags = android.content.Intent.FLAG_ACTIVITY_NEW_TASK
                activity.startActivity(intent)
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun invokeDartEvent(method: String, arguments: Any? = null) {
        mainHandler.post {
            methodChannel.invokeMethod(method, arguments)
        }
    }

    private fun startRecording(result: Result) {
        Log.d(TAG, "Flutter called startRecording")
        if (bgService == null) {
            result.error("NO_SERVICE", "Background service not bound yet", null)
            return
        }
        mainHandler.post {
            bgService?.cameraControl?.startRecording(object : CameraControl.RecordingCallback {
                override fun onSuccess() {
                    Log.d(TAG, "Recording started")
                    invokeDartEvent("onRecordingStarted")
                    result.success(null)
                }

                override fun onFailed(reason: String) {
                    Log.e(TAG, "Recording failed: $reason")
                    invokeDartEvent("onRecordingFailed", mapOf("reason" to reason))
                    result.error("FAILED", reason, null)
                }
            })
        }
    }

    private fun stopRecording(result: Result) {
        Log.d(TAG, "Flutter called stopRecording")
        if (bgService == null) {
            result.error("NO_SERVICE", "Background service not bound yet", null)
            return
        }
        mainHandler.post {
            bgService?.cameraControl?.stopRecording(object : CameraControl.RecordingCallback {
                override fun onSuccess() {
                    Log.d(TAG, "Recording stopped")
                    invokeDartEvent("onRecordingStopped")
                    result.success(null)
                }

                override fun onFailed(reason: String) {
                    Log.e(TAG, "Stop recording failed: $reason")
                    invokeDartEvent("onRecordingStopFailed", mapOf("reason" to reason))
                    result.error("FAILED", reason, null)
                }
            })
        }
    }

    private fun connectCamera(result: Result) {
        Log.d(TAG, "Flutter called connectCamera")
        mainHandler.post {
            bgService?.cameraControl?.connect()
            result.success(null)
        }
    }

    private fun scanWifi(result: Result) {
        Log.d(TAG, "Flutter called scanWifi")
        mainHandler.post {
            // 1. Immediately return cached results if any
            val cached = wifiHelper.getCachedScanResults()
            val cachedList = cached.map { 
                mapOf("ssid" to it.ssid, "level" to it.level)
            }
            if (cachedList.isNotEmpty()) {
                invokeDartEvent("onWifiList", cachedList)
            }

            // 2. Trigger a fresh scan in background
            wifiHelper.startScan() 
            
            result.success(null) // Return null, results come via onWifiList event
        }
    }

    private fun connectWifi(ssid: String, password: String, result: Result) {
        Log.d(TAG, "Flutter called connectWifi: $ssid")
        mainHandler.post {
            wifiHelper.connectToNetwork(ssid, password)
            result.success(null)
        }
    }

    // Called from init when wifiHelper.onConnectionStatus fires
    private fun onWifiConnected() {
        Log.d(TAG, "Wi-Fi connected! Now auto-connecting camera SDK...")
        mainHandler.post {
            bgService?.cameraControl?.connect()
        }
    }

    private fun saveRecordingMetadata(jsonStr: String, result: Result) {
        try {
            val dir = File(activity.filesDir, "recordings")
            if (!dir.exists()) dir.mkdirs()

            val file = File(dir, "recordings.json")
            val existingJson = if (file.exists()) file.readText() else "[]"
            val recordings = JSONArray(existingJson)

            recordings.put(JSONObject(jsonStr))
            file.writeText(recordings.toString(2))
            
            invokeDartEvent("onMetadataSaved")
            result.success(null)
        } catch (e: Exception) {
            val safeMsg = e.message ?: "Unknown error"
            invokeDartEvent("onMetadataSaveFailed", mapOf("reason" to safeMsg))
            result.error("FAILED", safeMsg, null)
        }
    }

    private fun getRecordings(): String {
        return try {
            val file = File(activity.filesDir, "recordings/recordings.json")
            if (file.exists()) file.readText() else "[]"
        } catch (e: Exception) {
            "[]"
        }
    }

    private fun setDemoMode(enabled: Boolean, result: Result) {
        demoMode = enabled
        result.success(null)
    }

    private fun exportRecording(recordingId: String, result: Result) {
        if (bgService == null) {
            result.error("NO_SERVICE", "Background service not bound yet", null)
            return
        }
        val filePaths = bgService?.cameraControl?.lastCapturedFilePaths
        if (filePaths == null || filePaths.isEmpty()) {
            val safeMsg = "No recording files found. Make sure to record first."
            invokeDartEvent("onExportFailed", mapOf("error" to safeMsg, "resolution" to "all"))
            result.error("FAILED", safeMsg, null)
            return
        }

        mainHandler.post {
            // Fix the "Array<out String>" variance issue by creating a typed copy
            val pathsArray = filePaths.map { it }.toTypedArray()
            bgService?.videoExporter?.exportBothResolutions(
                pathsArray,
                recordingId,
                object : VideoExporter.ExportCallback {
                    override fun onProgress(progress: Float, resolution: String) {
                        val pct = (progress * 100).toInt()
                        invokeDartEvent("onExportProgress", mapOf("progress" to pct, "resolution" to resolution))
                    }

                    override fun onSuccess(outputPath: String, resolution: String) {
                        invokeDartEvent("onExportSuccess", mapOf("path" to outputPath, "resolution" to resolution))
                        if (resolution == "1080P") {
                            result.success("Export Complete")
                        }
                    }

                    override fun onFailed(error: String, resolution: String) {
                        invokeDartEvent("onExportFailed", mapOf("error" to error, "resolution" to resolution))
                        if (resolution == "1080P") {
                            result.error("EXPORT_FAILED", error, null)
                        }
                    }
                }
            )
        }
    }

    // ===== GOOGLE DRIVE =====

    private fun signInToDrive(result: Result) {
        Log.d(TAG, "Flutter called signInToDrive")
        if (activity is MainActivity) {
            (activity as MainActivity).signInToDrive()
            result.success(null)
        } else {
            result.error("NO_ACTIVITY", "MainActivity required", null)
        }
    }

    private fun getDriveSignInStatus(): String {
        return if (activity is MainActivity) {
            (activity as MainActivity).getDriveSignInStatus() ?: ""
        } else ""
    }

    fun onDriveSignInResult(email: String?, error: String?) {
        if (email != null) {
            invokeDartEvent("onDriveSignInSuccess", mapOf("email" to email))
        } else {
            invokeDartEvent("onDriveSignInFailed", mapOf("error" to (error ?: "Unknown error")))
        }
    }

    private fun uploadToDrive(filePath: String, result: Result) {
        Log.d(TAG, "Flutter called uploadToDrive: $filePath")
        val file = java.io.File(filePath)
        if (!file.exists()) {
            invokeDartEvent("onDriveUploadFailed", mapOf("error" to "File does not exist: $filePath"))
            result.error("NOT_FOUND", "File not found", null)
            return
        }

        bgService?.driveUploader?.uploadFile(file, object : DriveUploader.UploadCallback {
            override fun onProgress(progress: Int) {
                invokeDartEvent("onDriveUploadProgress", mapOf("progress" to progress))
            }
            override fun onSuccess(fileId: String) {
                invokeDartEvent("onDriveUploadSuccess", mapOf("fileId" to fileId))
            }
            override fun onError(error: String) {
                invokeDartEvent("onDriveUploadFailed", mapOf("error" to error))
            }
            override fun onQueued(filePath: String) {
                invokeDartEvent("onDriveUploadQueued", mapOf("filePath" to filePath))
            }
        })
        result.success(null)
    }

    private fun uploadPendingFiles(result: Result) {
        Log.d(TAG, "Flutter called uploadPendingFiles")
        bgService?.driveUploader?.uploadPendingFiles(object : DriveUploader.UploadCallback {
            override fun onProgress(progress: Int) {
                invokeDartEvent("onDriveUploadProgress", mapOf("progress" to progress))
            }
            override fun onSuccess(fileId: String) {
                invokeDartEvent("onDriveUploadSuccess", mapOf("fileId" to fileId))
            }
            override fun onError(error: String) {
                invokeDartEvent("onDriveUploadFailed", mapOf("error" to error))
            }
            override fun onQueued(filePath: String) {
                invokeDartEvent("onDriveUploadQueued", mapOf("filePath" to filePath))
            }
        })
        result.success(null)
    }
}

File: insta360-bridge-flutter/ios/Runner/Insta360Channel.swift
import Flutter
import UIKit
import INSCameraSDK

class Insta360Channel: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?
    private let wifiHelper = WifiHelper.shared
    private let recordingManager = RecordingManager.shared
    private var demoMode = false
    
    
    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.noveloffice.insta360bridge/camera", binaryMessenger: registrar.messenger())
        let instance = Insta360Channel()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        instance.setupCallbacks()
        
        // Auto-connect to last-used camera on launch
        instance.wifiHelper.autoConnect()
    }
    
    private func setupCallbacks() {
        // Wire WifiHelper events to Flutter
        wifiHelper.onScanResults = { [weak self] results in
            self?.invokeDartEvent("onWifiList", arguments: results)
        }
        
        wifiHelper.onConnectionStatus = { [weak self] status in
            self?.invokeDartEvent("onWifiConnected", arguments: ["status": status])
            if status == "CONNECTED" {
                // For B-end SDK, we use setup() after Wi-Fi is connected
                print("[Insta360Channel] WiFi connected, triggering SDK camera setup")
                INSCameraManager.socket().setup()
            }
        }
    }
    
    // MARK: - Method Call Handler
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        // Camera control
        case "getStatus":
            result(getCameraStatus())
        case "connectCamera":
            INSCameraManager.socket().setup()
            result(nil)
        case "getExportedFiles":
            let files = getExportedFilesList()
            if let data = try? JSONSerialization.data(withJSONObject: files),
               let jsonString = String(data: data, encoding: .utf8) {
                result(jsonString)
            } else {
                result("[]")
            }
        case "getRecordings":
            result(recordingManager.getRecordings())
        case "startRecording":
            startRecording(result: result)
        case "stopRecording":
            stopRecording(result: result)
            
        // WiFi management
        case "scanWifi":
            wifiHelper.scanWifi()
            result(nil)
        case "stopScan":
            // On iOS, scanning is just returning a list, but we acknowledge the stop request
            print("[Insta360Channel] stopScan called")
            result(nil)
        case "connectWifi":
            let args = call.arguments as? [String: Any] ?? [:]
            let ssid = args["ssid"] as? String ?? ""
            let password = args["password"] as? String ?? ""
            wifiHelper.connectToNetwork(ssid: ssid, password: password)
            result(nil)
        case "getSavedCameras":
            let cameras = wifiHelper.getSavedCameras()
            if let data = try? JSONSerialization.data(withJSONObject: cameras),
               let jsonString = String(data: data, encoding: .utf8) {
                result(jsonString)
            } else {
                result("[]")
            }
        case "removeSavedCamera":
            let args = call.arguments as? [String: Any] ?? [:]
            let ssid = args["ssid"] as? String ?? ""
            wifiHelper.removeSavedCamera(ssid: ssid)
            result(nil)
        case "getCurrentSSID":
            print("[Insta360Channel] Fetching current SSID...")
            wifiHelper.fetchCurrentSSID { ssid in
                print("[Insta360Channel] SSID fetch complete: \(ssid ?? "nil")")
                result(ssid)
            }
        case "openWifiSettings":
            // Attempt multiple known schemes to get directly to the WIFI tab
            let schemes = ["App-Prefs:root=WIFI", "App-Prefs:WIFI", "prefs:root=WIFI"]
            for scheme in schemes {
                if let url = URL(string: scheme), UIApplication.shared.canOpenURL(url) {
                    UIApplication.shared.open(url, options: [:], completionHandler: nil)
                    result(nil)
                    return
                }
            }
            // Fallback to general settings
            if let settingsUrl = URL(string: UIApplication.openSettingsURLString) {
                UIApplication.shared.open(settingsUrl)
            }
            result(nil)
            
        // Recordings
        case "exportRecording":
            let args = call.arguments as? [String: Any] ?? [:]
            let recordingId = args["recordingId"] as? String ?? ""
            exportRecording(recordingId: recordingId, result: result)
            
        // Google Drive — placeholder
        case "signInToDrive":
            signInToDrive(result: result)
        case "uploadToDrive":
            let args = call.arguments as? [String: Any] ?? [:]
            let filePath = args["filePath"] as? String ?? ""
            uploadToDrive(filePath: filePath, result: result)
        case "uploadPendingFiles":
            result(FlutterError(code: "NOT_IMPLEMENTED", message: "iOS Google Drive is in development", details: nil))
        case "getPendingUploadCount":
            result(0)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - Send events to Flutter
    
    private func invokeDartEvent(_ method: String, arguments: Any? = nil) {
        DispatchQueue.main.async { [weak self] in
            self?.channel?.invokeMethod(method, arguments: arguments)
        }
    }
    
    // MARK: - Camera Control Implementation
    
    private func getCameraStatus() -> String {
        if demoMode { return "CONNECTED" }
        
        let state = INSCameraManager.socket().cameraState
        switch state {
        case .connected:
            return "CONNECTED"
        case .found, .synchronized:
            return "CONNECTING"
        case .noConnection, .connectFailed:
            return "DISCONNECTED"
        @unknown default:
            return "DISCONNECTED"
        }
    }
    
    private func startRecording(result: @escaping FlutterResult) {
        print("[Insta360Channel] startRecording called")
        if demoMode {
            invokeDartEvent("onRecordingStarted")
            result(nil)
            return
        }
        
        // Use INSCaptureOptions as required by the B-end SDK
        let options = INSCaptureOptions()
        
        // In the B-end SDK, capture methods are on the commandManager
        INSCameraManager.shared().commandManager.startCapture(with: options) { (error: Error?) in
            if let error = error {
                print("[Insta360Channel] startCapture error: \(error.localizedDescription)")
                DispatchQueue.main.async {
                    self.invokeDartEvent("onRecordingFailed", arguments: ["reason": error.localizedDescription])
                }
            } else {
                print("[Insta360Channel] startCapture success")
                DispatchQueue.main.async {
                    self.invokeDartEvent("onRecordingStarted")
                }
            }
        }
        result(nil)
    }
    
    private func exportRecording(recordingId: String, result: @escaping FlutterResult) {
        // Need to find the .insv files from the camera first
        INSCameraManager.shared().commandManager.fetchCameraFileList { (error, fileList) in
            guard let fileList = fileList else {
                self.invokeDartEvent("onExportFailed", arguments: ["error": "No files found on camera", "resolution": "all"])
                result(FlutterError(code: "FAILED", message: "No files found", details: nil))
                return
            }
            
            // For now, take the last one
            let lastFile = fileList.last as? String ?? ""
            let paths = [lastFile]
            
            VideoExporter.shared.exportBothResolutions(recordingId: recordingId, filePaths: paths) { progress, res in
                let pct = Int(progress * 100)
                self.invokeDartEvent("onExportProgress", arguments: ["progress": pct, "resolution": res])
            } completion: { path4K, path1080p, error in
                if let error = error {
                    self.invokeDartEvent("onExportFailed", arguments: ["error": error.localizedDescription, "resolution": "all"])
                    result(FlutterError(code: "FAILED", message: error.localizedDescription, details: nil))
                } else {
                    self.invokeDartEvent("onExportSuccess", arguments: ["path": path1080p ?? "", "resolution": "1080p"])
                    self.invokeDartEvent("onExportSuccess", arguments: ["path": path4K ?? "", "resolution": "4K"])
                    result("Export successful")
                }
            }
        }
    }
    
    private func signInToDrive(result: @escaping FlutterResult) {
        // AppAuth implementation goes here, for now use a placeholder
        print("[Insta360Channel] signInToDrive called")
        // Normally this involves opening a UI and calling DriveUploader.shared.saveAuthState()
        result(nil)
    }
    
    private func uploadToDrive(filePath: String, result: @escaping FlutterResult) {
        let url = URL(fileURLWithPath: filePath)
        DriveUploader.shared.upload(fileURL: url) { fileID, error in
            if let error = error {
                self.invokeDartEvent("onDriveUploadFailed", arguments: ["error": error.localizedDescription])
                result(FlutterError(code: "FAILED", message: error.localizedDescription, details: nil))
            } else {
                self.invokeDartEvent("onDriveUploadSuccess", arguments: ["fileID": fileID ?? ""])
                result(fileID)
            }
        }
    private func getExportedFilesList() -> [[String: Any]] {
        let documentsDir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        let dir = documentsDir.appendingPathComponent("exported_videos")
        
        var results = [[String: Any]]()
        
        let fileManager = FileManager.default
        guard let files = try? fileManager.contentsOfDirectory(at: dir, includingPropertiesForKeys: [.contentModificationDateKey, .fileSizeKey], options: .skippingHiddenFiles) else {
            return []
        }
        
        for fileURL in files where fileURL.pathExtension == "mp4" {
            let attr = try? fileManager.attributesOfItem(atPath: fileURL.path)
            let size = attr?[.size] as? Int64 ?? 0
            let date = attr?[.modificationDate] as? Date ?? Date()
            
            let resolution = fileURL.lastPathComponent.contains("_4K_") ? "4K (3840×1920)" : "1080p (1920×960)"
            
            results.append([
                "name": fileURL.lastPathComponent,
                "path": fileURL.path,
                "size": size,
                "sizeFormatted": formatFileSize(size),
                "resolution": resolution,
                "lastModified": Int64(date.timeIntervalSince1970 * 1000)
            ])
        }
        
        return results.sorted { ($0["lastModified"] as? Int64 ?? 0) > ($1["lastModified"] as? Int64 ?? 0) }
    }
    
    private func formatFileSize(_ bytes: Int64) -> String {
        let kb = Double(bytes) / 1024.0
        let mb = kb / 1024.0
        let gb = mb / 1024.0
        if gb >= 1 { return String(format: "%.1f GB", gb) }
        if mb >= 1 { return String(format: "%.1f MB", mb) }
        if kb >= 1 { return String(format: "%.1f KB", kb) }
        return "\(bytes) B"
    }
}
